### PR TITLE
feat: symfony 7 support

### LINKED ISSUE #2143: Bundle cannot be used without Doctrine Annotations
Symfony's FrameworkBundle automatically integrates Doctrine annotations by exposing an `annotation.reader` service that this bundle consumes. However, it is possible to disable that integration:

```YAML
framework:
    annotations: false
```

If I do that, the container won't compile anymore:

> The service "nelmio_api_doc.model_describers.object" has a dependency on a non-existent service "annotations.reader".

Modern Symfony 6 projects would rather use PHP attributes than Doctrine Annotations, so enabling the annotations integration isn't really necessary anymore. NelmioApiDocBundle could not use the annotations reader for anything in such a scenario because there are no annotations it could read from.

Symfony 6.4 triggers a deprecation if the annotations integration is enabled and Symfony 7 simply won't provide that service anymore.

NelmioApiDocBundle should continue to work if I disable the integration of Doctrine Annotations. In that case, Doctrine-style annotations would not be taken into account anymore. Furthermore, NelmioApiDocBundle should not list Doctrine Annotations as a mandatory dependency anymore. Applications that still use Doctrine Annotations should directly depend on that library instead.

### PR DESCRIPTION
**Pull request by faizanakram99**
- makes doctrine/annotations optional
- updates tests and ci to test with both symfony 7 and < symfony 7 version
- adds routes-attributes.yaml file for symfony 7 compatible controllers/routes
- updates some fixtures to use attributes when sf7 kernel is installed
- skips tests/packages related to packages like jms serializer, bazinga, etc not supported by symfony 7 for symfony 7 matrix

Fixes #2143 and #2147, supersedes #2151 and #2155

#SymfonyHackday

### CODE REVIEW

**Comment by faizanakram99** [comment]
Tests are failing with this error, i am not sure why

```
Multiple @OA\Response() with the same response="200":
  \Nelmio\ApiDocBundle\Tests\Functional\Controller\InvokableController->__invoke() in path/to/NelmioApiDocBundle/Tests/Functional/Controller/InvokableController.php
```

```
  \Nelmio\ApiDocBundle\Tests\Functional\Controller\SerializedNameController->serializedNameAction() in path/to/NelmioApiDocBundle/Tests/Functional/Controller/SerializedNameController.php" at DefaultLogger.php line 31
```

**Comment by faizanakram99** [comment]
Ok from the looks of it, i don't think it will be possible to support both sf 7 and previous versions, sf 7 conflicts with sensio framework extra and it is probably required for all previous sf versions.

Looks like we need a major version bump and drop all those packages altogether.

File: Tests/Functional/Entity/SymfonyConstraintsWithValidationGroups.php [hunk 1]
```diff
@@ -12,32 +12,64 @@
 namespace Nelmio\ApiDocBundle\Tests\Functional\Entity;
 
 use OpenApi\Annotations as OA;
+use OpenApi\Attributes\Items;
+use OpenApi\Attributes\Property;
```
Comment (DjordyKoert) [inline Tests/Functional/Entity/SymfonyConstraintsWithValidationGroups.php:16]: We could also keep the same use notation. Will leave this up to you :)
```php
use OpenApi\Attributes as OAT
```

File: Tests/Functional/TestKernel.php [hunk 2]
```diff
@@ -37,75 +37,79 @@
...
     public function __construct(int $flags = 0)
     {
         parent::__construct('test'.$flags, true);
 
         $this->flags = $flags;
     }
 
-    /**
-     * {@inheritdoc}
-     */
```
Comment (DjordyKoert) [inline Tests/Functional/TestKernel.php:58]: Why was this one removed?

File: Tests/Functional/TestKernel.php [hunk 3]
```diff
@@ -121,9 +125,6 @@ private function import($routes, $resource, $prefix, $type)
         }
     }
 
-    /**
-     * {@inheritdoc}
-     */
```
Comment (DjordyKoert) [inline Tests/Functional/TestKernel.php:126]: Same for this one

**Review by DjordyKoert — APPROVED** [review]
Thanks a lot :)

**Review by DjordyKoert — CHANGES_REQUESTED** [review]

File: Tests/Functional/TestKernel.php (hunk 3 above)
Comment (faizanakram99) [reply Tests/Functional/TestKernel.php:126]: It's probably cs fixer, I ran cs fix on whole Tests folder with `@Symfony` rules

Comment (DjordyKoert) [reply Tests/Functional/TestKernel.php:126]: Ah thats what I assumed as well, just making sure

File: .github/workflows/continuous-integration.yml [hunk 4]
```diff
@@ -59,13 +61,19 @@ jobs:
           key: ${{ runner.os }}-composer-${{ hashFiles('**/composer.json') }}
           restore-keys: ${{ runner.os }}-composer-
 
+      - name: Remove packages not compatible symfony 7
+        if: matrix.symfony-require == '7.0'
```
[the remark this answers is not in the document]
Comment (faizanakram99) [reply .github/workflows/continuous-integration.yml:65]: yes indeed

**Comment by RobertMe** [comment]
I find it quite rude to use mine (and @DjordyKoert) our changes and commit them under your own name. Especially since at least my PR is waiting for review, and not because I abondended it (/not because didn't incorperate requested changes).

Your could have just merged both branches / pull requests into your own branch so it would be clear that someone else wrote the code, and add your own changes on top of it.

File: Tests/Functional/Entity/SymfonyConstraintsWithValidationGroups.php (hunk 1 above)
Comment (faizanakram99) [reply Tests/Functional/Entity/SymfonyConstraintsWithValidationGroups.php]: done

File: .github/workflows/continuous-integration.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (faizanakram99) [reply .github/workflows/continuous-integration.yml]: fixed

**Comment by faizanakram99** [comment]
> I find it quite rude to use mine (and @DjordyKoert) our changes and commit them under your own name. Especially since at least my PR is waiting for review, and not because I abondended it (/not because didn't incorperate requested changes).
> 
> Your could have just merged both branches / pull requests into your own branch so it would be clear that someone else wrote the code, and add your own changes on top of it.

apologies, i did mention it in a comment that I missed that there was a PR https://github.com/nelmio/NelmioApiDocBundle/pull/2151#issuecomment-1849886108

Later I tried to fork your repositories so that I add my changes on top of it but since I had already made all the changes locally (including both making doctrine/annotations optional and adding sf 7 support), i found it hard to separate them and apply on top of your changes. If you wish, I can give you write permissions to this repository (actually you and @DjordyKoert both), so that you may re-add your commits or even ammend existing commits.

It doesn't really matter to me whom the commits are attributed to, I want support for sf7 in this lib because we are using it  and it is blocking our sf 7 upgrade and it seemed a good task for symfony hack day. Again apology for any miscommunications.

**Comment by DjordyKoert** [comment]
> I find it quite rude to use mine (and @DjordyKoert) our changes and commit them under your own name. Especially since at least my PR is waiting for review, and not because I abondended it (/not because didn't incorperate requested changes).
> 
> Your could have just merged both branches / pull requests into your own branch so it would be clear that someone else wrote the code, and add your own changes on top of it.

I personally have no issue with my 'changes'/commits being lost, since my PR contained the bare minimum (and didn't even work to begin with). I am just very happy @faizanakram99 was able to make it (almost? because CI is still failing) work for Symfony 7. 😄

**Comment by DjordyKoert** [comment]
> Tests are failing with this error, i am not sure why
> 
> ```
> Multiple @OA\Response() with the same response="200":
>   \Nelmio\ApiDocBundle\Tests\Functional\Controller\InvokableController->__invoke() in path/to/NelmioApiDocBundle/Tests/Functional/Controller/InvokableController.php
> ```
> 
> ```
>   \Nelmio\ApiDocBundle\Tests\Functional\Controller\SerializedNameController->serializedNameAction() in path/to/NelmioApiDocBundle/Tests/Functional/Controller/SerializedNameController.php" at DefaultLogger.php line 31
> ```

I can also reproduce this locally. This is a very weird one since it's appending the response from `InvokableController` to the `SerializedNameController` `/api/serializename` path?????

**Comment by DjordyKoert** [comment]
@faizanakram99 I have created a PR on your fork repo with changes to make phpunit work for Symfony 7. https://github.com/faizanakram99/NelmioApiDocBundle/pull/1

Older versions are still failing but this should be a good starting point for fixing the others 😄

**Comment by faizanakram99** [comment]
> @faizanakram99 I have created a PR on your fork repo with changes to make phpunit work for Symfony 7. [faizanakram99#1](https://github.com/faizanakram99/NelmioApiDocBundle/pull/1)
> 
> Older versions are still failing but this should be a good starting point for fixing the others 😄

Thank you so much @DjordyKoert , I've merged your changes (with rebase)

I think you also need to run cs-fix on changed files, style ci is complaining

File: .github/workflows/continuous-integration.yml [hunk 5]
```diff
@@ -59,13 +61,19 @@ jobs:
           key: ${{ runner.os }}-composer-${{ hashFiles('**/composer.json') }}
           restore-keys: ${{ runner.os }}-composer-
 
+      - name: Remove packages not compatible symfony 7
+        if: matrix.symfony-require == '7.0.*'
+        run: |
+          composer remove friendsofsymfony/rest-bundle sensio/framework-extra-bundle jms/serializer-bundle willdurand/hateoas-bundle --no-update --dev
```
Comment (derrabus) [inline .github/workflows/continuous-integration.yml:67]: `jms/serializer-bundle` and `willdurand/hateoas-bundle` have been fixed in the meantime.

**Comment by DjordyKoert** [comment]
@faizanakram99 https://github.com/faizanakram99/NelmioApiDocBundle/pull/2

I finally managed to fix each and every failing test (minus 7.2 https://github.com/nelmio/NelmioApiDocBundle/pull/2162 fixes this) (this was quite the task 😬). Could you kindly merge this into this PR. 

After this is done we simply need to wait for one of the maintainers to merge this and then Nelmio will officially have Symfony 7 support! 🚀

**Comment by DjordyKoert** [comment]
@faizanakram99 could you please run a style fix 😄

**Review by DjordyKoert — APPROVED** [review]

**Comment by DjordyKoert** [comment]
> @faizanakram99 could you please run a style fix 😄, mine doesn't really fix anything for some reason

Nevermind that, I simply applied the diff from styleci https://github.com/faizanakram99/NelmioApiDocBundle/pull/3 😃 

@faizanakram99 a merge and styleci should be fixed

**Comment by DjordyKoert** [comment]
@GuilhemN any possibility you can look at this? ❤️

File: .github/workflows/continuous-integration.yml [hunk 6]
```diff
@@ -59,13 +67,24 @@ jobs:
...
+          composer remove friendsofsymfony/rest-bundle sensio/framework-extra-bundle --no-update --dev
+
       - name: "Install dependencies with composer"
         env:
           SYMFONY_REQUIRE: "${{ matrix.symfony-require }}"
         run: |
           composer global config --no-plugins allow-plugins.symfony/flex true
           composer global require --no-progress --no-scripts --no-plugins symfony/flex
           composer update --no-interaction --no-progress ${{ matrix.composer-flags }}
+          composer update --no-interaction --no-progress ${{ matrix.composer-flags }}
```
Comment (fabianoroberto) [inline .github/workflows/continuous-integration.yml:87]: Why this line appear twice `composer update --no-interaction --no-progress ${{ matrix.composer-flags }}`

**Comment by Gemorroj** [comment]
@GuilhemN ping

**Comment by GuilhemN** [comment]
Thank you very much for the huge work!
I will try to look into fixing the CI a bit later.

Happy new year, best wishes!

**Comment by DjordyKoert** [comment]
Thank you very much @GuilhemN !

Happy new years & best wishes to you too :heart:

**Pull request merged**